index/segment/mem/segment.go

```diff
@@ -0,0 +1,258 @@
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package mem
+
+import (
+	"errors"
+	"regexp"
+	"sync"
+
+	"github.com/m3db/m3ninx/doc"
+	"github.com/m3db/m3ninx/index"
+	sgmt "github.com/m3db/m3ninx/index/segment"
+	"github.com/m3db/m3ninx/postings"
+)
+
+var (
+	errSegmentClosed     = errors.New("segment is closed")
+	errSegmentSealed     = errors.New("segment has been sealed")
+	errUnknownPostingsID = errors.New("unknown postings ID specified")
+)
+
+type segment struct {
+	opts   Options
+	offset int
+
+	wg sync.WaitGroup
+
+	// Internal state of the segment. The allowed transitions are:
+	//   - Open -> Sealed -> Closed
+	//   - Open -> Closed
+	state struct {
+		sync.RWMutex
+
+		sealed bool
+		closed bool
+	}
+
+	// Mapping of postings list ID to document.
+	docs struct {
+		sync.RWMutex
+
+		data []doc.Document
+	}
+
+	// Current writer and reader IDs. Writers increment the writer ID for each new
+	// document and only increment the reader ID after the document has been fully
+	// indexed by the segment. Readers do not need to acquire a lock.
+	ids struct {
+		sync.RWMutex
+
+		writer, reader postings.AtomicID
+	}
+
+	// Mapping of field (name and value) to postings list.
+	termsDict termsDict
+}
+
+// NewSegment returns a new in-memory mutable segment. It will start assigning
+// postings IDs at offset+1.
+func NewSegment(offset postings.ID, opts Options) (sgmt.MutableSegment, error) {
+	s := &segment{
+		opts:      opts,
+		offset:    int(offset) + 1, // The first ID assigned by the segment is offset + 1.
+		termsDict: newSimpleTermsDict(opts),
+	}
+
+	s.docs.data = make([]doc.Document, opts.InitialCapacity())
+
+	s.ids.writer = postings.NewAtomicID(offset)
+	s.ids.reader = postings.NewAtomicID(offset)
+	return s, nil
+}
+
+func (s *segment) Insert(d doc.Document) error {
+	s.state.RLock()
+	if s.state.closed {
+		s.state.RUnlock()
+		return errSegmentClosed
+	}
+
+	if s.state.sealed {
+		s.state.RUnlock()
+		return errSegmentSealed
+	}
+
+	// TODO: Consider supporting concurrent writes by relaxing the requirement that
+	// inserted documents are immediately searchable.
+	s.ids.Lock()
+
+	newID := s.ids.writer.Inc()
+	s.insertDoc(newID, d)
+	err := s.insertTerms(newID, d)
+	s.ids.reader.Inc()
+
+	s.ids.Unlock()
+
+	s.state.RUnlock()
+	return err
+}
+
+func (s *segment) Reader() (index.Reader, error) {
+	s.state.RLock()
+	if s.state.closed {
+		s.state.RUnlock()
+		return nil, errSegmentClosed
+	}
+
+	maxID := s.ids.reader.Load()
+	r := newReader(s, maxID, &s.wg)
```

Comment:
ah you're using the wg for ref counting? how about using https://github.com/m3db/m3x/blob/master/checked/types.go#L81 instead?

Reply:
We can use `Ref`, but it seems a bit heavyweight at the moment so I think it might be simpler to stick with just the wait group for now.